lib/spack/spack/build_environment.py

```diff
@@ -1014,26 +1014,33 @@ def add_modifications_for_dep(dep):
 
 
 def get_cmake_prefix_path(pkg):
-    # Note that unlike modifications_from_dependencies, this does not include
-    # any edits to CMAKE_PREFIX_PATH defined in custom
-    # setup_dependent_build_environment implementations of dependency packages
     build_deps      = set(pkg.spec.dependencies(deptype=('build', 'test')))
     link_deps       = set(pkg.spec.traverse(root=False, deptype=('link')))
     build_link_deps = build_deps | link_deps
     spack_built = []
     externals = []
+
     # modifications_from_dependencies updates CMAKE_PREFIX_PATH by first
     # prepending all externals and then all non-externals
     for dspec in pkg.spec.traverse(root=False, order='post'):
         if dspec in build_link_deps:
             if dspec.external:
-                externals.insert(0, dspec)
+                externals.insert(0, dspec.prefix)
             else:
-                spack_built.insert(0, dspec)
-
-    ordered_build_link_deps = spack_built + externals
-    build_link_prefixes = filter_system_paths(
-        x.prefix for x in ordered_build_link_deps)
+                spack_built.insert(0, dspec.prefix)
+
+    # CMAKE_PREFIX_PATH here cannot come from the outside environment under
+    # normal circumstances, it has to come from spack.  include these second
+    # after the spack_built dependencies but before externals
+    env_mod_list = modifications_from_dependencies(pkg.spec, 'build',
+                                                   custom_mods_only=True)
+    env_mod = {}
+    env_mod_list.apply_modifications(env_mod)
+    extra_prefix_paths = env_mod.get("CMAKE_PREFIX_PATH", "").split(";")
+
+    ordered_build_link_deps = spack_built + extra_prefix_paths + externals
```

Comment:
This generally preserves the desired order, although I don't know if I have a concise way to describe what exactly that order is. My best summary is:

* Spack-built prefixes (i.e. the install prefix of each Spack-built package)
* Custom modifications from the DAG, with priority given to spack-built-packages (this is nebulous since those methods could prepend, append, or even override entirely, so It might be more accurate that spack-built packages have the "last say")
* External prefixes

Overall I agree that is fine. If you think it is useful, then perhaps it's worth adding a comment describing this ordering to the method's docstring.

Reply:
I'll add a comment soon